WalletWasabi/Blockchain/TransactionOutputs/CoinsRegistry.cs

```diff
@@ -230,57 +148,128 @@ public bool IsKnown(uint256 txid)
 	{
 		lock (Lock)
 		{
-			return KnownTransactions.Contains(txid);
+			return CoinsByTransactionsIds.ContainsKey(txid);
 		}
 	}
 
-	public bool TryGetSpenderSmartCoinsByOutPoint(OutPoint outPoint, [NotNullWhen(true)] out HashSet<SmartCoin>? coins)
+	public bool TryGetCoinsByInputPrevOut(OutPoint outPoint, [NotNullWhen(true)] out HashSet<SmartCoin>? coins)
 	{
 		lock (Lock)
 		{
-			return TryGetSpenderSmartCoinsByOutPointNoLock(outPoint, out coins);
+			return TryGetCoinsByInputPrevOutNoLock(outPoint, out coins);
 		}
 	}
 
-	private bool TryGetSpenderSmartCoinsByOutPointNoLock(OutPoint outPoint, [NotNullWhen(true)] out HashSet<SmartCoin>? coins)
+	private bool TryGetCoinsByInputPrevOutNoLock(OutPoint outPoint, [NotNullWhen(true)] out HashSet<SmartCoin>? coins)
 	{
-		return CoinsByOutPoint.TryGetValue(outPoint, out coins);
+		coins = null;
+		lock (Lock)
+		{
+			if (!TxIdsByInputsPrevOut.TryGetValue(outPoint, out var txId))
+			{
+				return false;
+			}
+
+			if (!CoinsByTransactionsIds.TryGetValue(txId, out var coinsByOutpoints))
+			{
+				return false;
+			}
+
+			coins = coinsByOutpoints.Values.ToHashSet();
+		}
+		return true;
 	}
 
-	public bool TryGetSpentCoinByOutPoint(OutPoint outPoint, [NotNullWhen(true)] out SmartCoin? coin)
+	public bool TryGetCoinByOutPoint(OutPoint outPoint, [NotNullWhen(true)] out SmartCoin? coin)
 	{
+		coin = null;
 		lock (Lock)
 		{
-			return SpentCoinsByOutPoint.TryGetValue(outPoint, out coin);
+			if (!CoinsByTransactionsIds.TryGetValue(outPoint.Hash, out var coinsByOutpoint))
+			{
+				return false;
+			}
+
+			if (!coinsByOutpoint.TryGetValue(outPoint, out coin))
+			{
+				return false;
+			}
 		}
+
+		return true;
 	}
 
 	internal (ICoinsView toRemove, ICoinsView toAdd) Undo(uint256 txId)
 	{
 		lock (Lock)
 		{
 			var allCoins = AsAllCoinsViewNoLock();
-			var toRemove = new List<SmartCoin>();
-			var toAdd = new List<SmartCoin>();
+			var toRemove = new HashSet<SmartCoin>();
+			var toAdd = new HashSet<SmartCoin>();
```

Comment:
```suggestion
			HashSet<SmartCoin> toRemove = new();
			HashSet<SmartCoin> toAdd = new();
```
so we eliminate nullability?